### PR TITLE
Upgrade Django Select 2

### LINKED ISSUE #1792: Update select2.multi-checkboxes.js to work with django-select2 7.2.0
* https://github.com/applegrew/django-select2/releases
* https://github.com/wasikuss/select2-multi-checkboxes
* https://github.com/select2/select2/releases

### PR DESCRIPTION
**Pull request by vaibhavmule**
fix #1792

Upgraded `django_select2` from `7.1.1` to `7.2.2` thus `select2js` upgraded from `3.5.x` to `4.x`.

Select2 4.0 introduced [AMD-based build](https://addyosmani.com/writing-modular-js/) so I have generated `select2.multi-checkboxes.js` from `amd` branch of [wasikuss/select2-multi-checkboxes](https://github.com/wasikuss/select2-multi-checkboxes/tree/amd).

`Select2 4.x` doesn't have `.wrap` class around dropdown text so I needed to change CSS in  hypha/static_src/src/sass/apply/components/_select2.scss.  

@frjo I wasn't sure if I need to change CSS for https://github.com/OpenTechFund/hypha/blob/master/hypha/static_src/src/sass/public/components/_select2.scss as I have gone through the code and I haven't found where it is used.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -25,7 +25,7 @@ django-tables2==1.21.2
 django-tinymce4-lite==1.7.5
 django-two-factor-auth==1.9.1
 django-webpack-loader==0.6.0
-django_select2==7.1.1
+django_select2==7.2.0
```
Comment (frjo) [inline requirements.txt]: I see that 7.2.1 is the latest now.

File: hypha/static_src/src/sass/apply/components/_select2.scss [hunk 2]
```diff
@@ -68,30 +68,24 @@ $dropdown-height: 45px;
...
+        &[aria-selected='true'] {
             &::before {
-                min-width: 20px;
-                height: 20px;
-                margin-right: 10px;
-                background: $color--white;
-                border: 1px solid $color--mid-grey;
+                background: url('./../../images/tick.svg') $color--dark-blue center no-repeat;
+                background-size: 12px;
+                border: 1px solid $color--dark-blue;
```
Comment (frjo) [inline hypha/static_src/src/sass/apply/components/_select2.scss]: Please make the same changes in the public version of this file. Not used I think but best to keep synced.

File: hypha/apply/dashboard/static/js/select2.multi-checkboxes.js [hunk 3]
```diff
@@ -3,77 +3,124 @@
...
-    var self = this;
-    self.options = options;
-    self.$element = $(element);
-  var values = self.$element.val();
-    self.$element.removeAttr('multiple');
-    self.select2 = self.$element.select2({
-      allowClear: true,
-      minimumResultsForSearch: options.minimumResultsForSearch,
-      placeholder: options.placeholder,
-      closeOnSelect: false,
```
Comment (frjo) [inline hypha/apply/dashboard/static/js/select2.multi-checkboxes.js]: The select popup get closed as soon as you make a selection, this did not happen before and this setting might be what fixed that.

File: hypha/apply/dashboard/static/js/django_select2-checkboxes.js [hunk 4]
```diff
@@ -1,23 +1,34 @@
...
+
+      $(function () {
         $('.django-select2-checkboxes').each(function (i, element) {
-            var $element = $(element);
-            init($element);
+          var $element = $(element);
+
+          $element.select2({
+            placeholder: $element.data('placeholder'),
+            templateSelection: function(data) {
```
Comment (frjo) [inline hypha/apply/dashboard/static/js/django_select2-checkboxes.js]: Tried to add `closeOnSelect: false,` and it seems to work but then I get some errors when closing the popup.

`TypeError: undefined is not an object (evaluating 'r.GetData(this,"element").select2')` in `select2.min.js`

`TypeError: undefined is not an object (evaluating 'match[0].classList')` in `/apply/submission-filters.js:41`

File: requirements.txt (hunk 1 above)
Comment (vaibhavmule) [reply requirements.txt]: It is not yet released at pip https://pypi.org/project/django-select2/#history

File: hypha/static_src/src/sass/apply/components/_select2.scss (hunk 2 above)
Comment (vaibhavmule) [reply hypha/static_src/src/sass/apply/components/_select2.scss]: Done

File: hypha/apply/dashboard/static/js/django_select2-checkboxes.js (hunk 4 above)
Comment (vaibhavmule) [reply hypha/apply/dashboard/static/js/django_select2-checkboxes.js]: Let me check this.

Comment (vaibhavmule) [reply hypha/apply/dashboard/static/js/django_select2-checkboxes.js]: yes `closeOnSelect` works but give errors on the console.
I tried looking for what exactly the issue is, this is what I have got but that didn't seem to help as there is no resolution on it.
 https://github.com/select2/select2/issues/3282
https://github.com/select2/select2/issues/3800
I will spend some time more and see if I can have a workaround, I tried the following. 
```
          if ($element !== undefined) {
            $element.select2({
              placeholder: $element.data('placeholder'),
              closeOnSelect: false,
              templateSelection: function(data) {
                let filterType = $element.data('placeholder');

                if (!data.selected.length) {
                  return filterType
                } else if (data.selected.length == data.all.length) {
                  return 'All ' + filterType + ' selected';
                }
                return data.selected.length + ' of ' + data.all.length + ' ' + filterType + ' selected';
              },
              selectionAdapter: SelectionAdapter,
              returnesultsAdapter: ResultsAdapter
            });
          };
```

Comment (vaibhavmule) [reply hypha/apply/dashboard/static/js/django_select2-checkboxes.js]: @frjo 

The thing is jQuery and $ were available when it applies select2 checkbox plugin for the first time but not afterward, 

There are solutions like changing sequence of jquery and other plugin used in HTML file, but that breaks more things.

File: hypha/apply/dashboard/static/js/select2.multi-checkboxes.js (hunk 3 above)
Comment (vaibhavmule) [reply hypha/apply/dashboard/static/js/select2.multi-checkboxes.js]: This was kind of overwriting closeOnSelect options but in the latest version, it is amd based build and everything is moduler, so select-multi-checkboxes just gives you utils not options you have to pass yourself.

File: requirements.txt (hunk 1 above)
Comment (vaibhavmule) [reply requirements.txt]: I have raised issue on the repo, most probably it will get fixed. https://github.com/applegrew/django-select2/issues/594

**Comment by frjo** [comment]
Thanks for fixing and investigating on this. Will be on the next deploy.

File: requirements.txt (hunk 1 above)
Comment (vaibhavmule) [reply requirements.txt]: @frjo 
The issue is resolved django-select2 released new version 7.2.2.
I have made changes in a branch to use `django-select2=7.2.2`

**Pull request merged**